File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) {
        if (lists.length === 0) return null;

        for (let i = 0; i < lists.length; i++) {
            lists[i] = this.internalMergeFn(lists[i], lists[i - 1]);
        }

        return lists[lists.length - 1];
    }

    internalMergeFn(l1, l2) {
        const copy = new ListNode()
        let end = copy;

        while (l1 && l2) {
            if (l1.val < l2.val) {
                end.next = l1;
                l1 = l1.next;
            } else {
                end.next = l2;
                l2 = l2.next;
            }
            end = end.next;
        }
        if (l1) {
            end.next = l1;
        }
        if (l2) {
            end.next = l2;
        }
        return copy.next;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int len = 0;
        ListNode current = head;

        while (current != null) {
            current = current.next;
            len += 1;
        }

        int targetIndex = len - n;
        if (targetIndex == 0) return head.next;

        current = head;
        for (int i = 0;i<len; ++i) {
            if ((i + 1) == targetIndex) {
                current.next = current.next.next;
                break;
            }
            current = current.next;
        }
        return head;
    }
}
